src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesImplementation.cs

```diff
@@ -39,28 +43,35 @@ public sealed override void GenerateKey()
 
         protected sealed override void Dispose(bool disposing)
         {
+            InvalidateOneShotAlgorithms();
             base.Dispose(disposing);
         }
 
+        public override byte[] Key
+        {
+            set
+            {
+                base.Key = value;
+                InvalidateOneShotAlgorithms();
```

Comment:
* This instance has been used, so there's a cached lite-cipher for the mode we care about.
* Thread A calls into set_Key, hasn't yet made it to updating the field
* Thread B starts encrypting (or decrypting, whatever), something long.  Gigabyteish.  Still has k0
* Thread A makes it out to InvalidateOneShotAlgorithms.  get_Key will now return k1
* Thread B finishes.  The field is null, so it puts back the instance it used at k0.
* As long as parallel ops never happen again, the cache stays at k0 from one-borrow one-return.

Right?

Reply:
Ah, right, yep.